File: python/diarization/embedding.py
from __future__ import annotations

import threading
from pathlib import Path
from typing import TYPE_CHECKING

import numpy as np

from .fbank import FBank

if TYPE_CHECKING:
    import onnxruntime as ort

SAMPLE_RATE = 16_000
EMBEDDING_DIM = 256
FRAME_LENGTH_SAMPLES = 400
FRAME_SHIFT_SAMPLES = 160
NUM_MEL_BINS = 80
MIN_INPUT_SAMPLES = 16_000

DEFAULT_INPUT_NAME = "feats"
DEFAULT_OUTPUT_NAME = "embs"

class EmbeddingModel:
    def __init__(
        self,
        session: "ort.InferenceSession",
        input_name: str = DEFAULT_INPUT_NAME,
        output_name: str = DEFAULT_OUTPUT_NAME,
    ):
        self._session = session
        self._lock = threading.Lock()
        self._input_name = input_name
        self._output_name = output_name
        self._fbank = FBank(NUM_MEL_BINS, FRAME_LENGTH_SAMPLES, FRAME_SHIFT_SAMPLES)

    @classmethod
    def load(
        cls,
        model_path: str | Path,
        input_name: str = DEFAULT_INPUT_NAME,
        output_name: str = DEFAULT_OUTPUT_NAME,
        providers: list[str] | None = None,
    ) -> EmbeddingModel:
        import onnxruntime as ort
        path = Path(model_path)
        sess_options = ort.SessionOptions()
        sess_options.graph_optimization_level = ort.GraphOptimizationLevel.ORT_ENABLE_ALL
        sess_options.intra_op_num_threads = 1
        session = ort.InferenceSession(
            str(path),
            sess_options=sess_options,
            providers=providers or ["CPUExecutionProvider"],
        )
        return cls(session, input_name, output_name)

    @property
    def sample_rate(self) -> int:
        return SAMPLE_RATE

    @property
    def embedding_dim(self) -> int:
        return EMBEDDING_DIM

    @property
    def min_input_samples(self) -> int:
        return MIN_INPUT_SAMPLES

    def embed(self, samples: np.ndarray) -> np.ndarray:
        if samples.shape[0] < FRAME_LENGTH_SAMPLES:
            raise ValueError(
                f"embedding: input {samples.shape[0]} samples shorter than "
                f"frame length {FRAME_LENGTH_SAMPLES}"
            )
        feats_flat = self._fbank.compute(samples)
        frames = feats_flat.size // NUM_MEL_BINS
        feats = feats_flat.reshape(1, frames, NUM_MEL_BINS).astype(np.float32, copy=False)
        with self._lock:
            outputs = self._session.run(None, {self._input_name: feats})
        emb = outputs[0]
        if emb.size == 0 or emb.shape[-1] != EMBEDDING_DIM:
            raise ValueError(
                f"embedding: expected last dim {EMBEDDING_DIM}, got shape {emb.shape}"
            )
        flat = emb.astype(np.float32, copy=False).reshape(-1)
        return _l2_normalize(flat)

def _l2_normalize(v: np.ndarray) -> np.ndarray:
    norm = float(np.linalg.norm(v))
    if norm < 1e-9:
        norm = 1e-9
    return v / norm

def cosine_sim(a: np.ndarray, b: np.ndarray) -> float:
    if a.shape != b.shape:
        raise ValueError(f"cosine_sim: shape mismatch {a.shape} vs {b.shape}")
    return float(np.dot(a, b))

File: go/internal/diarization/embedding.go
package diarization

import (
	"fmt"
	"math"
	"sync"

	ort "github.com/yalue/onnxruntime_go"
)

const (
	EmbeddingSampleRate         uint32 = 16000
	EmbeddingDim                       = 256
	EmbeddingFrameLengthSamples        = 400
	EmbeddingFrameShiftSamples         = 160
	EmbeddingNumMelBins                = 80
	EmbeddingMinInputSamples           = 16000

	embeddingInputName  = "feats"
	embeddingOutputName = "embs"
)

type EmbeddingModel struct {
	mu      sync.Mutex
	session *ort.DynamicAdvancedSession
	fbank   *FBank
}

func LoadEmbedding(modelPath string) (*EmbeddingModel, error) {
	return LoadEmbeddingWithIONames(modelPath, embeddingInputName, embeddingOutputName)
}

func LoadEmbeddingWithIONames(modelPath, inputName, outputName string) (*EmbeddingModel, error) {
	if modelPath == "" {
		return nil, fmt.Errorf("embedding: empty model path")
	}
	if err := initOnnxRuntime(); err != nil {
		return nil, fmt.Errorf("embedding: init onnxruntime: %w", err)
	}
	sess, err := ort.NewDynamicAdvancedSession(
		modelPath,
		[]string{inputName},
		[]string{outputName},
		nil,
	)
	if err != nil {
		return nil, fmt.Errorf("embedding: load %q: %w", modelPath, err)
	}
	return &EmbeddingModel{
		session: sess,
		fbank:   NewFBank(EmbeddingNumMelBins, EmbeddingFrameLengthSamples, EmbeddingFrameShiftSamples),
	}, nil
}

func (m *EmbeddingModel) SampleRate() uint32   { return EmbeddingSampleRate }
func (m *EmbeddingModel) Dim() int             { return EmbeddingDim }
func (m *EmbeddingModel) MinInputSamples() int { return EmbeddingMinInputSamples }

func (m *EmbeddingModel) Close() error {
	m.mu.Lock()
	defer m.mu.Unlock()
	if m.session != nil {
		m.session.Destroy()
		m.session = nil
	}
	return nil
}

func (m *EmbeddingModel) Embed(samples []float32) ([]float32, error) {
	if len(samples) < EmbeddingFrameLengthSamples {
		return nil, fmt.Errorf("embedding: input %d samples shorter than frame length %d",
			len(samples), EmbeddingFrameLengthSamples)
	}

	feats, err := m.fbank.Compute(samples)
	if err != nil {
		return nil, err
	}
	frames := len(feats) / EmbeddingNumMelBins

	in, err := ort.NewTensor(ort.NewShape(1, int64(frames), int64(EmbeddingNumMelBins)), feats)
	if err != nil {
		return nil, fmt.Errorf("embedding: input tensor: %w", err)
	}
	defer in.Destroy()

	m.mu.Lock()
	defer m.mu.Unlock()
	if m.session == nil {
		return nil, fmt.Errorf("embedding: session closed")
	}

	outputs := []ort.Value{nil}
	if err := m.session.Run([]ort.Value{in}, outputs); err != nil {
		return nil, fmt.Errorf("embedding: run: %w", err)
	}
	defer outputs[0].Destroy()

	tensor, ok := outputs[0].(*ort.Tensor[float32])
	if !ok {
		return nil, fmt.Errorf("embedding: unexpected output type %T", outputs[0])
	}
	shape := tensor.GetShape()
	if len(shape) == 0 || int(shape[len(shape)-1]) != EmbeddingDim {
		return nil, fmt.Errorf("embedding: expected last dim %d, got shape %v", EmbeddingDim, shape)
	}
	src := tensor.GetData()
	out := make([]float32, len(src))
	copy(out, src)
	l2Normalize(out)
	return out, nil
}

func l2Normalize(v []float32) {
	var sum float32
	for _, x := range v {
		sum += x * x
	}
	norm := float32(math.Sqrt(float64(sum)))
	if norm < normFloor {
		norm = normFloor
	}
	inv := 1 / norm
	for i := range v {
		v[i] *= inv
	}
}
